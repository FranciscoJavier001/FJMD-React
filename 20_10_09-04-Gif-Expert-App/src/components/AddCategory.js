import React, { useState } from 'react'
import PropTypes from 'prop-types';

export const AddCategory = ({setCategories}) => { //** Agregamos el AddCategory pero va directamente al setCategories y va a volver a renderizar, y se recibe props */

    const [inputValue, setInputValue] = useState('') /** Estas son las dos funciones que tiene el inputvalue (donde escribimos) y abajo ya quedo declarado  */

    const handleInputChange = (e) => { 
        setInputValue(e.target.value) /** Esto me permite cambiar el texto */
    }

    const handleSubmit = (e) => { /** Aqui hacemos la peticion para que salga en el DOM, y que se vea reflejado cuando damos enter  */
        e.preventDefault(); /** Esto evita el refresh de la pagina */

        if(inputValue.trim().length > 0){ //** Solo si el arreglo es mayor de 2 que empieze a buscar */
            //** Asi le hizimos para que la primera saliera arriba */
            setCategories(cats => [inputValue, ...cats,]); //** El primer callback es el estado anterior y luego regresa un nuevo estado con todos los nuevos arreglos, se llama a esta instruccion para tener acceso al arreglo, porque se esta pasando como argumento */
            setInputValue('');
        }
    }

    return ( /** Aqui directamente ya estamos en el DOM */
        // El Form agrupa todo lo demas, por eso no hace falta el fragment
        //** Aqui le decimo que va a responder a handleSubmit, osea cuando de enter */
        <form onSubmit={handleSubmit}>  
            <input 
                type="text"
                value={inputValue}
                onChange={handleInputChange} /** Aqui le doy el valor para cambiarlo */
                />
            </form>
    )
}

AddCategory.prototype = { /** Para que utilicen nuestro componente con el peque√±o candado con los argumentos que estoy esperando, esta es la peticion de los proptypes  */
    setCategories: PropTypes.func.isRequired
}