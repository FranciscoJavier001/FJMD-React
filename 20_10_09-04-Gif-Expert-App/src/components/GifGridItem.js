import React from 'react'

export const GifGridItem = ({title, url}) => { //** Estos son los parametros que se reciben en el map del GifGrid.js */

    return ( //** Es lo que retorna el arreglo de GifGridItem */
        // Estos son los estilos de bootstrap, y aqui vemos que solo se hace el llamado de los parametros dichos en el getGifs.js
        // Como lo vez aqui el div con el className se nos puso a modo de bootstrap, le ponemos el nombre, que en este caso es card y luego la animamos como queramos
        <div className="card animate__animated animate__fadeIn"> 
            {/* Estos son los parametros que trae cada imagen */}
            <img src={url} alt={title}/> 
            {/* Es el titulo que va abajo */}
            <p>{title}</p>
        </div>
    )
}
