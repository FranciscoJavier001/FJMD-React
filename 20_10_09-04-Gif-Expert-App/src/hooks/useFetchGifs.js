//** use = Es un Hook, y no es mas que una funcion  */
import { useEffect, useState } from "react"
import {getGifs} from '../helpers/getGifs'

export const useFetchGifs = (category) => { //** Esta va a ser la peticion automatica cuando se carge porque cambie la categoria */
    const [state, setState] = useState({ //** El useState le dice a los componentes que algo cambio y deben renderizarse */
        data: [], //** Este es su estado inicial */
        loading: true //** Este es su estado inicial */
    });

    //** Esto retorna una promesa por lo cual se pone un then, para tener las imgs y la funcion de fecha y llamamos al setImagenes  */
    useEffect (() => { //** Este lo utilizamos para que no se haga un llamado de todas al agregar una nueva categoria, recibe la funcion que quiero ejecutar  */
        getGifs(category) //** Aqui se hace el llamado a la funcion que queremos */
        .then(imgs => {

                setState({ //** Este codigo significa que solo quiero que se renderize la primera vez que se llama, ya despues no */
                    data: imgs,
                    loading: false
                });
            })
    }, [category]) //** Aqui es un arreglo de dependencias, y solo se va a disparar una unica vez, y se la categoria cambia, entonces solo va a ejecutar otra vez este useEffect */

    return state; //{data:[], loading: true}; //** El state es el estado del useState, y cuando cambia es lo que me retorna */
}
