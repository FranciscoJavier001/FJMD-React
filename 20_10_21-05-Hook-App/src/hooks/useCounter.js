//** La falla en este archivo es que lo usamos en MultipleCustomHooks y cambiamos los nombres, cambiamos de state por counter y el setState por setCounter (antes115)*/

// rafc nos crea un functional component
import { useState } from 'react'; //** Siempre hay que importar un estado, pero este es una funcion de react que esta definida, que significa, dale estas caracteristicas a un objeto */

export const useCounter = ( initialState = 10 ) => { //** Publica useCounter con un initialState con valor de 10, pero que esta declarado en 1 , em layout */
    
    const [counter, setCounter] = useState(initialState); // 10 //** Pon atencion en el CounterWithCustomHook, con el state, recuerda el nombre que tiene para poderlo usar desde aqui que aqui importamos las funciones */

    const reset = () => { //** Funcion de flecha que no hace nada, solo llama al setCounter */
        setCounter( initialState ); //** Aqui mandamos el valor del initialState que es de 10 */
    }

    const increment = () => { //** Funcion exportada al LayOut */
        setCounter( counter + 1 );
    }

    const decrement = () => {
        setCounter( counter - 1 );
    }

    return { //** Mandamos la referencia a las funciones */
        counter, //** Esta definido en CounterWithCustomHook, en el LayOut y aqui */
        increment,
        decrement,
        reset
    };
}