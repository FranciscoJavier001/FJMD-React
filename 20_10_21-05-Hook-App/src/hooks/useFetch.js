import { useState, useEffect, useRef } from 'react';

export const useFetch = ( url ) => { //** Aqui pedimos el url, y cuando lo recibimos que haga algo */
    
    const isMounted = useRef(true);
    const [state, setState] = useState({data: null, loading: true, error: null}); //** Este es el estado del state */

    useEffect( () => { //**  */
        return () => {
            isMounted.current = false;
        }
    }, [])


    useEffect( () => { //** Cuando cambie algo */

        setState({ data: null, loading: true, error: null });

        fetch( url ) //** Madamos el url */
            .then( resp => resp.json() ) //** Responde con una promesa y vamos a extraer el json (que se convierte en un string) */
            .then( data => { //** Esto devuelve otra promesa donde ya tenemos la data, y esta data es la respuesta del endPoint */

                if ( isMounted.current ) {
                    setState({ //** Ahora esto se va a cambiar cuando reciba algo */
                        loading: false,
                        error: null,
                        data
                    });
                }

            })
            .catch( () => {
                setState({
                    data: null,
                    loading: false,
                    error: 'No se pudo cargar la info'
                })
            })

    },[url])

    return state; //** Retornamos el state y esta regresando solo el objeto, y ahora solo esta regresando el state */
}
