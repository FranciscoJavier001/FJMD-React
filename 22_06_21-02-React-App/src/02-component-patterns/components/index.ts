//** Para no tener un monton de importaciones */

export { ProductButtons } from './ProductButtons';
export { ProductCard } from './ProductCard';
export { ProductImage } from './ProductImage';
export { ProductTitle } from './ProductTitle';