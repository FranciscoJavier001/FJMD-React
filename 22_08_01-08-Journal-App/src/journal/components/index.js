export * from './Navbar';
export * from './SideBar';
export * from './ImageGallery';