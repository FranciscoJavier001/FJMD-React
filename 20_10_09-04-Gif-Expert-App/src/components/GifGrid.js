import React  from 'react'
import { useFetchGifs } from '../hooks/useFetchGifs'
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({ category }) => { //** Aqui lo exportamos para que lo pueda utilizar GifExpertApp, el argumento es category */

    const {data:images, loading} = useFetchGifs(category); //** Es la peticion http para traer las imagenes que corresponden a la categoria, con los : se renombra en la desestructuracion */

    return (
        // Cuando haya que regresar 2 objetos usa el fragment
        <>
            {/* Este h3 son donde se quedan guardadanas nuestras busquedas */}
            <h3 className="card animate__animated animate__fadeIn">{category}</h3>

            {/* Esta es cuando se estan cargando */}
            {loading && <p className="card animate__animated flash">Loading</p>}

            <div className="card-grid">
                {
                    images.map(img => ( //** Asi hago un return de objetos */
                        <GifGridItem //** Aqui se va a mostrar */
                            key={img.id} //** Este es el return ue se me hizo, de id y de img */
                            {...img} //** Operador spred de img  */
                        />
                    ))
                }
            </div>
        </>
    )
}
