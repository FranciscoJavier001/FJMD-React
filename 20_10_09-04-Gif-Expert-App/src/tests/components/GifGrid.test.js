import React from 'react' //** Importamos React para las pruebas */
import '@testing-library/jest-dom' //** Para evitar errores de escritura */
import {shallow} from 'enzyme'  //** Importamos Shallow desde enzyme */
import { GifGrid } from '../../components/GifGrid' //** Importamos el componente al cual le vamos a hacr la prueba */
import { useFetchGifs } from '../../hooks/useFetchGifs';
jest.mock('../../hooks/useFetchGifs') //** El mook finge una llamada al archivo y controlar la informacion que va a responder */

describe('Pruebas en el <GifGrid/>', () => {
    
    const category = "Death Note"; //** category esta definida como requerida, por eso le ponemos algo para las pruebas */
    
    test('Debe de mostrarse corectamente', () => {

        useFetchGifs.mockReturnValue({ //** En las llaves va el objeto que va a tomar el componente del useFetchGifs en el GifGrid */
            data: [], //** Estos son los valores que el useFetchGifs regresa */
            loading: true
        }) //** Vamos a hacer una falsa implementacion */

        const wrapper = shallow(<GifGrid category={category}/>) //** El wrapper es por default con el shallos y pasandole el componente de cual es el sujeto de prueba y pasandole el argumento category porque es requerido */
        
        expect(wrapper).toMatchSnapshot(); //** Esto hace el snapshot */
    })

    // Deben de ser pruebas unitarias, debemos de probar los componentes de manera aislada

    test('Debe de mostrar items cuando se cargan imagenes con useFetchGifs', () => {
        //** Hacemos un mook que es commo finguir algo, para que el componente crea que tiene informacion con la data del customeHook, es decir como si ya me hubieran regresado la informacion cuando ya la cargo */

        const gifs = [{ //** Esta es la informacion que le vamos a pasar al arreglo de abajo */
            id: 'ABC',
            title: 'CualquierCosa',
            url: 'https://localhost/cualquier/cosa.jpg'
        },
    {
            id: '123',
            title: 'CualquierCosa',
            url: 'https://localhost/cualquier/cosa.jpg'
    }]

        useFetchGifs.mockReturnValue({ 
            data: gifs, //** La data van a ser los gifs */
            loading: false //** Esto es para decir que el loading ya se cargo  */
        })

        const wrapper = shallow(<GifGrid category={category}/>) //** Renderizamos el mismo componente, porque no lo pusimos arriba */

        expect(wrapper).toMatchSnapshot(); //** Es para crear el snapshot */
        expect(wrapper.find('p').exists()).toBe(false) //** Si el parrado eiste, yo esperaria que fuera false */
        expect(wrapper.find('GifGridItem').length).toBe(gifs.length)//** Asi seleccionamos un elemento en el GifGrid.test.js.snap, el length para saber cuantos gifs tengo */
    })
})