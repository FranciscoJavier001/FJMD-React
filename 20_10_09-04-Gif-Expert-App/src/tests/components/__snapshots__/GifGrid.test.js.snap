// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <GifGrid/> Debe de mostrar items cuando se cargan imagenes con useFetchGifs 1`] = `
<Fragment>
  <h3
    className="card animate__animated animate__fadeIn"
  >
    Death Note
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="ABC"
      key="ABC"
      title="CualquierCosa"
      url="https://localhost/cualquier/cosa.jpg"
    />
    <GifGridItem
      id="123"
      key="123"
      title="CualquierCosa"
      url="https://localhost/cualquier/cosa.jpg"
    />
  </div>
</Fragment>
`;

exports[`Pruebas en el <GifGrid/> Debe de mostrarse corectamente 1`] = `
<Fragment>
  <h3
    className="card animate__animated animate__fadeIn"
  >
    Death Note
  </h3>
  <p
    className="card animate__animated flash"
  >
    Loading
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
